glusterd2/servers/sunrpc/handshake_prog.go

```diff
@@ -140,41 +140,66 @@ func (p *GfHandshake) ServerGetspec(args *GfGetspecReq, reply *GfGetspecRsp) err
 		).Error("ServerGetspec(): failed to read volfile")
 		goto Out
 	}
+	reply.Spec = string(content)
 
 	// If Volfile not available in volfiles directory
 	// fetch from etcd store
 	if os.IsNotExist(err) {
 		// Reset error due to Volfile not exists
 		err = nil
 
-		resp, err := store.Get(context.TODO(), volfilePrefix+volfileID)
+		if strings.HasPrefix(volfileID, "snaps/") {
+			snapname := strings.Replace(volfileID, "snaps/", "", -1)
+			snapvol, err := snapshot.GetSnapshot(snapname)
+			if err != nil {
+				log.WithError(err).WithField(
+					"volfile", volfileID,
+				).Error("failed to get snapshot info")
+				goto Out
+			}
+			volinfo = &snapvol.SnapVolinfo
+		} else {
+			volinfo, err = volume.GetVolume(volfileID)
+			if err != nil {
+				log.WithError(err).WithField(
+					"volfile", volfileID,
+				).Error("failed to get volume info")
+				goto Out
+			}
+		}
+
+		tmpl, err := volgen.GetTemplateFromVolinfo(volinfo, "client")
```

Comment:
is there any performance import if we generate volfiles on demand?

Reply:
https://github.com/gluster/glusterd2/pull/1363#issuecomment-444382322